app/src/modules/trigger/trigger.c

```diff
@@ -9,101 +9,372 @@
+	}
+}
+
+/* STATE_NORMAL_MODE */
+
+static void normal_mode_init(void *o)
+{
+	struct s_object *user_object = o;
+
+	LOG_DBG("normal_mode_init");
+
+	LOG_DBG("Setting trigger work timeout to %lld seconds", user_object->update_interval_sec);
+
+	/* Set update trigger timeout to the configured interval */
+	user_object->update_interval_current_sec = user_object->update_interval_sec;
+
+	k_work_reschedule(&trigger_work, K_SECONDS(user_object->update_interval_current_sec));
+}
+
+static void normal_mode_run(void *o)
+{
+	struct s_object *user_object = o;
+
+	LOG_DBG("normal_mode_run");
+
+	if (user_object->chan == &BUTTON_CHAN) {
+		LOG_DBG("Button %d pressed in normal mode, going into RTM",
+			user_object->button_number);
+
+		smf_set_state(SMF_CTX(&state_object), &states[STATE_REAL_TIME_MODE]);
+	} else if (user_object->chan == &CONFIG_CHAN) {
+		LOG_DBG("Configuration received in normal mode, going into RTM");
+
+		smf_set_state(SMF_CTX(&state_object), &states[STATE_REAL_TIME_MODE]);
+	}
+}
+
+/* STATE_DISCONNECTED */
+
+static void disconnected_run(void *o)
+{
+	ARG_UNUSED(o);
+
+	LOG_DBG("disconnected_run");
+
+	struct s_object *user_object = o;
+
+	if ((user_object->status == CLOUD_CONNECTED) && user_object->chan == &CLOUD_CHAN) {
+		smf_set_state(SMF_CTX(&state_object), &states[STATE_CONNECTED]);
+		return;
 	}
 }
 
+/* Construct state table */
+static const struct smf_state states[] = {
+	[STATE_INIT] = SMF_CREATE_STATE(
+		init_entry, init_run, NULL, NULL, NULL
+	),
+	[STATE_CONNECTED] = SMF_CREATE_STATE(
+		NULL, connected_run, connected_exit, NULL, &states[STATE_REAL_TIME_MODE]
+	),
+	[STATE_REAL_TIME_MODE] = SMF_CREATE_STATE(
+		real_time_mode_init, real_time_mode_run, NULL, &states[STATE_CONNECTED], NULL
+	),
+	[STATE_NORMAL_MODE] = SMF_CREATE_STATE(
+		normal_mode_init, normal_mode_run, NULL, &states[STATE_CONNECTED], NULL
+	),
+	[STATE_DISCONNECTED] = SMF_CREATE_STATE(
+		NULL, disconnected_run, NULL, NULL, NULL
+	)
+};
+
+/* Function called when there is a message received on a channel that the module listens to */
 void trigger_callback(const struct zbus_channel *chan)
 {
+	int err;
+
+	if ((chan != &CONFIG_CHAN) &&
+	    (chan != &CLOUD_CHAN) &&
+	    (chan != &BUTTON_CHAN) &&
+	    (chan != &PRIV_TIMER_CHAN)) {
+		LOG_ERR("Unknown channel");
+		return;
+	}
+
+	/* Update the state object with the channel that the message was received on */
+	state_object.chan = chan;
+
+	/* Copy corresponding data to the state object depending on the incoming channel */
 	if (&CONFIG_CHAN == chan) {
-		/* Get update interval configuration from channel. */
 		const struct configuration *config = zbus_chan_const_msg(chan);
 
 		if (config->config_present == false) {
 			LOG_DBG("Configuration not present");
 			return;
 		}
 
-		LOG_DBG("New update interval: %lld", config->update_interval);
+		state_object.update_interval_sec = config->update_interval;
+	}
 
-		update_interval = K_SECONDS(config->update_interval);
+	if (&CLOUD_CHAN == chan) {
```

Comment:
I've been thinking of these checks that we need to do with channel check and casting to the right types.
I put them in the state handlers in transport module, but not sure that's the right way. Could we macrofy it in a good and readable way?